File: src/api-service/__app__/onefuzzlib/notifications/teams.py
#!/usr/bin/env python
#
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.

import logging
from typing import Any, Dict, List, Optional, Union

import requests
from onefuzztypes.models import RegressionReport, Report, TeamsTemplate
from onefuzztypes.primitives import Container

from ..azure.containers import auth_download_url
from ..secrets import get_secret_string_value
from ..tasks.config import get_setup_container
from ..tasks.main import Task


def markdown_escape(data: str) -> str:
    values = r"\\*_{}[]()#+-.!"  # noqa: P103
    for value in values:
        data = data.replace(value, "\\" + value)
    data = data.replace("`", "``")
    return data


def code_block(data: str) -> str:
    data = data.replace("`", "``")
    return "\n```\n%s\n```\n" % data


def send_teams_webhook(
    config: TeamsTemplate,
    title: str,
    facts: List[Dict[str, str]],
    text: Optional[str],
) -> None:
    title = markdown_escape(title)

    message: Dict[str, Any] = {
        "@type": "MessageCard",
        "@context": "https://schema.org/extensions",
        "summary": title,
        "sections": [{"activityTitle": title, "facts": facts}],
    }

    if text:
        message["sections"].append({"text": text})

    config_url = get_secret_string_value(config.url)
    response = requests.post(config_url, json=message)
    if not response.ok:
        logging.error("webhook failed %s %s", response.status_code, response.content)


def notify_teams(
    config: TeamsTemplate,
    container: Container,
    filename: str,
    report: Optional[Union[Report, RegressionReport]],
) -> None:
    text = None
    facts: List[Dict[str, str]] = []

    if isinstance(report, Report):
        task = Task.get(report.job_id, report.task_id)
        if not task:
            logging.error(
                "report with invalid task %s:%s", report.job_id, report.task_id
            )
            return

        title = "new crash in %s: %s @ %s" % (
            report.executable,
            report.crash_type,
            report.crash_site,
        )

        links = [
            "[report](%s)" % auth_download_url(container, filename),
        ]

        setup_container = get_setup_container(task.config)
        if setup_container:
            links.append(
                "[executable](%s)"
                % auth_download_url(
                    setup_container,
                    report.executable.replace("setup/", "", 1),
                ),
            )

        if report.input_blob:
            links.append(
                "[input](%s)"
                % auth_download_url(
                    report.input_blob.container, report.input_blob.name
                ),
            )

        facts += [
            {"name": "Files", "value": " | ".join(links)},
            {
                "name": "Task",
                "value": markdown_escape(
                    "job_id: %s task_id: %s" % (report.job_id, report.task_id)
                ),
            },
            {
                "name": "Repro",
                "value": code_block(
                    "onefuzz repro create_and_connect %s %s" % (container, filename)
                ),
            },
        ]

        text = "## Call Stack\n" + "\n".join(code_block(x) for x in report.call_stack)

    else:
        title = "new file found"
        facts += [
            {
                "name": "file",
                "value": "[%s/%s](%s)"
                % (
                    markdown_escape(container),
                    markdown_escape(filename),
                    auth_download_url(container, filename),
                ),
            }
        ]

    send_teams_webhook(config, title, facts, text)

File: src/ApiService/ApiService/onefuzzlib/notifications/Teams.cs
using System.Net.Http;
using System.Text.Json;
using Microsoft.Extensions.Logging;
namespace Microsoft.OneFuzz.Service;

public interface ITeams {
    Async.Task NotifyTeams(TeamsTemplate config, Container container, IReport reportOrRegression, Guid notificationId);
}

public class Teams : ITeams {
    private readonly ILogger _logTracer;
    private readonly IOnefuzzContext _context;
    private readonly IHttpClientFactory _httpFactory;

    public Teams(IHttpClientFactory httpFactory, ILogger<Teams> logTracer, IOnefuzzContext context) {
        _logTracer = logTracer;
        _context = context;
        _httpFactory = httpFactory;
    }

    private static string CodeBlock(string data) {
        data = data.Replace("`", "``");
        return $"\n```\n{data}\n```\n";
    }

    private async Async.Task SendTeamsWebhook(TeamsTemplate config, string title, IList<Dictionary<string, string>> facts, string? text, Guid notificationId) {
        title = MarkdownEscape(title);

        var sections = new List<Dictionary<string, object>>() {
            new() {
                {"activityTitle", title},
                {"facts", facts}
            }
        };
        if (text != null) {
            sections.Add(new() {
                { "text", text }
            });
        }

        var message = new Dictionary<string, object>() {
            {"@type", "MessageCard"},
            {"@context", "https://schema.org/extensions"},
            {"summary", title},
            {"sections", sections}
        };

        var configUrl = await _context.SecretsOperations.GetSecretValue(config.Url.Secret);
        var client = new Request(_httpFactory.CreateClient());
        var response = await client.Post(url: new Uri(configUrl!), JsonSerializer.Serialize(message));
        if (response == null || !response.IsSuccessStatusCode) {
            _logTracer.LogError("webhook failed {NotificationId} {StatusCode} {content}", notificationId, response?.StatusCode, response?.Content);
        }
    }

    public async Async.Task NotifyTeams(TeamsTemplate config, Container container, IReport reportOrRegression, Guid notificationId) {
        var facts = new List<Dictionary<string, string>>();
        string? text = null;
        var title = string.Empty;
        var filename = reportOrRegression.FileName();

        if (reportOrRegression is Report report) {
            var task = await _context.TaskOperations.GetByJobIdAndTaskId(report.JobId, report.TaskId);
            if (task == null) {
                _logTracer.LogError("report with invalid task {JobId}:{TaskId}", report.JobId, report.TaskId);
                return;
            }

            title = $"new crash in {report.Executable}: {report.CrashType} @ {report.CrashSite}";

            var links = new List<string> {
                $"[report]({_context.Containers.AuthDownloadUrl(container, filename)})"
            };

            var setupContainer = Scheduler.GetSetupContainer(task.Config);
            if (setupContainer != null) {
                var setupFileName = NotificationsBase.ReplaceFirstSetup(report.Executable);
                links.Add(
                    $"[executable]({_context.Containers.AuthDownloadUrl(setupContainer, setupFileName)})"
                );
            }

            if (report.InputBlob != null) {
                links.Add(
                    $"[input]({_context.Containers.AuthDownloadUrl(report.InputBlob.Container, report.InputBlob.Name)})"
                );
            }

            facts.AddRange(new List<Dictionary<string, string>> {
                new() {{"name", "Files"}, {"value", string.Join(" | ", links)}},
                new() {
                    {"name", "Task"},
                    {"value", MarkdownEscape(
                        $"job_id: {report.JobId} task_id: {report.TaskId}"
                    )}
                },
                new() {
                    {"name", "Repro"},
                    {"value", CodeBlock($"onefuzz repro create_and_connect {container} {filename}")}
                }
            });

            text = "## Call Stack\n" + string.Join("\n", report.CallStack.Select(cs => CodeBlock(cs)));
        } else {
            title = "new file found";
            var fileUrl = _context.Containers.AuthDownloadUrl(container, filename);

            facts.Add(new Dictionary<string, string>() {
                {"name", "file"},
                {"value", $"[{MarkdownEscape(container.String)}/{MarkdownEscape(filename)}]({fileUrl})"}
            });
        }

        await SendTeamsWebhook(config, title, facts, text, notificationId);
    }

    private static string MarkdownEscape(string data) {
        var values = "\\*_{}[]()#+-.!";
        foreach (var c in values) {
            data = data.Replace(c.ToString(), "\\" + c);
        }
        data = data.Replace("`", "``");
        return data;
    }
}
